### PR TITLE
ARROW-446: [Python] Expand Sphinx documentation for 0.3

### PR DESCRIPTION
**Pull request by wesm**
I am going to finish the data model section and revamp the Parquet section, so we can get this pushed out with the release announcement tomorrow. We should continue to add a lot of new documentation over the coming weeks

### CODE REVIEW

File: python/doc/source/memory.rst [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+   buf = mmap = file_obj = None
+   !rm example.dat
+   !rm example2.dat
+
+In-Memory Reading and Writing
+-----------------------------
+
+To assist with serialization and deserialization of in-memory data, we have
+file interfaces that and read and write to Arrow Buffers.
```
Comment (jreback) [inline python/doc/source/memory.rst]: can

**Comment by wesm** [comment]
@xhochy I changed away from the RTD Sphinx theme -- I find the code examples, particularly string outputs, to be difficult to read. We can experiment with different themes

File: python/doc/source/memory.rst [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+* :class:`~pyarrow.HdfsFile`, for reading and writing data to the Hadoop Filesystem
+* :class:`~pyarrow.PythonFile`, for interfacing with Python file objects in C++
+
+We will discuss these in the following sections after explaining memory pools.
+
+Memory Pools
+------------
+
+All memory allocations and deallocations (like ``malloc`` and ``free`` in C)
+are tracked in an instance of ``arrow::MemoryPool``. This means that we can the
```
Comment (jreback) [inline python/doc/source/memory.rst]: then

**Comment by wesm** [comment]
Sounds like this is a known issue https://github.com/rtfd/sphinx_rtd_theme/issues/166 -- if we like the RTD theme we'll need to do some customization to improve the code highlighting to improve readability

**Comment by wesm** [comment]
I'm going to merge as soon as the build passes so I can get the website updated for the release announcement tomorrow

**Comment by wesm** [comment]
Can handle more edits and improvements in follow up patches

**Comment by wesm** [comment]
+1

**Comment by xhochy** [comment]
I recently favor more https://pypi.python.org/pypi/sphinx-bootstrap-theme/ (e.g. https://ryan-roemer.github.io/django-cloud-browser/ ) We could probably even customise it so that it looks more like the arrow homepage.

**Comment by wesm** [comment]
Sounds fine to me

**Pull request closed without merging**